doc/develop/languages/rust/index.rst

```diff
@@ -0,0 +1,163 @@
+   = note: the `thumbv7m-none-eabi` target may not be installed
+   = help: consider downloading the target with `rustup target add thumb7m-none-eabi`
+
+In this case, the given ``rustup`` command will install the needed
+target support.  The target needed will depend on both the board
+selected, as well as certain configuration choices (such as whether
+floating point is enabled).
+
+Writing a Rust Application
+**************************
+
+See :zephyr_file:`samples/rust` for examples of an application.
+
+CMake files
+-----------
+
+The application should contain a :file:`CMakeFiles.txt`, similar to
+the following:
+
+.. code-block:: cmake
+
+   cmake_minimum_required(VERSION 3.20.0)
+
+   find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+   project(hello_world)
+
+   rust_cargo_application()
+
+Cargo files
+-----------
+
+Rust applications are built with Cargo.  The Zephyr build system will
+configure and invoke cargo in your application directory, with options
+set so that it can find the Zephyr support libraries, and that the
+output will be contained within the Zephyr build directory.
+
+The :file:`Cargo.toml` will need to have a ``[lib]`` section that sets
+``crate-type = ["staticlib"]``, and will need to include ``zephyr =
+"0.1.0"`` as a dependency.  You can use crates.io and the Crate
+ecosystem to include any other dependencies you need.  Just make sure
+that you use crates that support building with no-std.
+
+Application
+-----------
+
+The application source itself should live in file:`src/lib.rs`.  A few
+things are needed.  A minimal file would be:
+
+.. code-block:: rust
+
+   #![no_std]
+
+   extern crate zephyr;
+
+   #[no_mangle]
+   extern "C" fn rust_main() {
+   }
+
+The ``no_std`` declaration is needed to prevent the code from
+referencing the ``std`` library.  The extern reference will cause the
+zephyr crate to be brought in, even if nothing from it is used.
+Practically, any meaningful Rust application on Zephyr will use
+something from this crate, and this line is not necessary.  Lastly,
+the main declaration exports the main symbol so that it can be called
+by C code.  The build ``rust_cargo_application()`` cmake function will
+include a small C file that will call into this from the C main
+function.
+
+Zephyr Functionality
+********************
+
+The bindings to Zephyr for Rust are under development, and are
+currently rather minimalistic.
+
+Bool Kconfig settings
+---------------------
+
+Boolean Kconfig settings can be used from within Rust code.  Due to
+design constraints by the Rust language, settings that affect
+compilation must be determined before the build is made.  In order to
+use this in your application, you will need to use the
+``zephyr-build`` crate, provided, to make these symbols available.
+
+To your ``Cargo.toml`` file, add the following:
+
+.. code-block:: toml
+
+   [build-dependencies]
+   zephyr-build = "0.1.0"
+
+Then, you will need a ``build.rs`` file to call the support function.
```

Comment:
I am not sure how I feel about using `build.rs` here. While it is fine, this is a cargo specific thing that any other build system with Rust support doesn't really like.
While there are a lot of crates that use it, I would prefer to avoid it if possible. Have you looked at the Linux kernel implementation for Kconfig in Rust?

Reply:
I _really_ want to make rust applications fully able to work with Cargo. Maybe there are other situations (such as adding code within Zephyr in Rust) where calling rustc directly would be useful. But for an application, where someone is quite likely to want external dependencies anyway, trying to do everything that cargo does is needlessly complicated.

Also, this is an almost trivial example of the kind of stuff we will need at build time to make Rust useful within Zephyr. In my other PR that brings in device tree, I'm also generating a devicetree.rs from the device tree for a particular build. It needs to run at build time.

I'm kind of trying to understand where the fear of cargo is coming from. Along with the memory issues and such, cargo and the crates community is one of the big reasons for someone to want to use Rust.

What I'm doing here is _very_ different from how Rust is being used in Linux. This is not about extending Zephyr with Rust, but about writing applications. It's more akin to writing Rust programs in Linux userspace.

Reply:
The problem with `build.rs` is that it is cargo specific rather than Rust specific. There are some other build systems that support Rust [(meson)](https://mesonbuild.com/), but they do not support `build.rs`.

Reply:
I’m not sure what you are looking for here, supporting application development with rust cargo makes immense sense.

Reply:
Yes, it makes sense to use cargo for application development, but by tying the devicetree and other buildtime stuff that Rust support needs to `build.rs` might have some unintended consequences:
1. The current devicetree implementation will not be compatible with writing drivers and zephyr internals which I imagine would not go the cargo route.
2. Make it impossible to use other build systems for applications since they do not support `build.rs`.

Note: I do not have any problem with running a rust file at build time. I just do not want to tie the build time generation we need to do to cargo.

Reply:
I'm not exactly sure what you would propose. `build.rs` is _how_ you do build-time extensions with Cargo. The file is very short, and calls code in `zephyr-build`. The bindgen code is more complex, but still can be handled whever.

I don't want to actively go against people that want to build Rust apps without using cargo, but I'm still a bit baffled as to why someone would want to do that. Cargo does a lot of stuff for us, and for the app developer, and it would be rather silly to try and reproduce all of that in Cmake. If Meson has replicated some of that work, then there would be a need to port that effort to Meson.

We will have many additional issues to address to incorporate rust code into Zephyr rather than just to use it, the least of which is getting a lot more broad approval of Rust.